# AI School - Epoch 1 - Checkpoint 2

**Total Points:** 100 points (Main exercises) + 10 bonus points

**Minimum Required:** You need **85/100 points** from the main exercises.

You don't need to complete everything perfectly. The bonus exercises are **optional but recommended** for extra practice with advanced topics.

**Dataset:** `omc_members.csv` : a synthetic (messy) dataset of Open Minds Club member records.

**Submission:** Create a GitHub repo, push your `.ipynb` file and the CSV, submit the repo link in the following form: https://forms.gle/h7Xqa78SHZiPXuM47

*Attempt the bonus if you finish early, and don't stress about perfection - learning is the goal!*


---

## Setup

In [1]:
import pandas as pd
import numpy as np

---

## Part 1: Data Loading & First Inspection

### Exo 1: Load the Data (5 points)

Load `omc_members.csv` into a DataFrame called `df`.

Then display:
- The **shape** of the DataFrame
- The **column names**
- The **data types** of each column


In [2]:
# Load the CSV
df = pd.read_csv('omc_members.csv')

In [3]:
# Display shape, column names, and dtypes
print("Shape:", df.shape)
print("Column Names:", df.columns.tolist())
print("Data Types:\n", df.dtypes)


Shape: (128, 14)
Column Names: ['member_id', 'full_name', 'age', 'wilaya', 'university', 'level', 'track', 'role', 'join_date', 'workshops_attended', 'project_score', 'github_profile', 'status', 'email']
Data Types:
 member_id                 str
full_name                 str
age                   float64
wilaya                    str
university                str
level                     str
track                     str
role                      str
join_date                 str
workshops_attended      int64
project_score         float64
github_profile            str
status                    str
email                     str
dtype: object


---

### Exo 2: First Look (5 points)

Display:
- The **first 5 rows** of the DataFrame
- A **statistical summary** of the numeric columns
- Use `.info()` to get a concise overview


In [4]:
# First 5 rows
print(df.head())



      member_id        full_name   age  wilaya    university level  \
0  OMC-2024-465  Nadine Bensalem  25.0   Sétif           ESI    L1   
1  OMC-2024-622      Imane Saidi  26.0   BLIDA          UMBB    M2   
2  OMC-2024-398    Sofiane Ziani  23.0  Béjaïa  Univ. Oran 1    M2   
3  OMC-2024-766     Wafa Rahmani  23.0  Annaba         ENSIA    M2   
4  OMC-2024-345   Nassim Rahmani  25.0   BATNA         ENSIA    M2   

                  track       role   join_date  workshops_attended  \
0       Web Development  Core Team  2024-04-14                   7   
1             AI & Data     Member  2024-07-04                   4   
2         Cybersecurity     Member  2024-02-27                   7   
3  Software Engineering  Core Team  2024-09-09                   2   
4       Web Development     Member  2024-07-19                   0   

   project_score        github_profile    status                      email  
0            NaN   github.com/nadine62    Active  nadine.bensalem@email.com  
1 

In [5]:
# Statistical summary
print(df.describe())



              age  workshops_attended  project_score
count  128.000000          128.000000     116.000000
mean    30.304688            3.500000      13.957759
std     87.099324            2.459114       3.433699
min      0.000000            0.000000       8.000000
25%     20.000000            1.000000      11.300000
50%     21.500000            3.000000      13.350000
75%     25.000000            6.000000      17.200000
max    999.000000            7.000000      20.000000


In [6]:
# Concise overview
print(df.info())


<class 'pandas.DataFrame'>
RangeIndex: 128 entries, 0 to 127
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   member_id           128 non-null    str    
 1   full_name           128 non-null    str    
 2   age                 128 non-null    float64
 3   wilaya              120 non-null    str    
 4   university          122 non-null    str    
 5   level               128 non-null    str    
 6   track               128 non-null    str    
 7   role                128 non-null    str    
 8   join_date           128 non-null    str    
 9   workshops_attended  128 non-null    int64  
 10  project_score       116 non-null    float64
 11  github_profile      113 non-null    str    
 12  status              128 non-null    str    
 13  email               128 non-null    str    
dtypes: float64(2), int64(1), str(11)
memory usage: 14.1 KB
None


---

### Exo 3: Missing Values Report (10 points)

1. Display the **count** of missing values per column
2. Display the **percentage** of missing values per column (rounded to 2 decimal places)
3. Print the names of columns that have **at least one** missing value


In [7]:
# Count of missing values per column
print(df.isnull().sum())


member_id              0
full_name              0
age                    0
wilaya                 8
university             6
level                  0
track                  0
role                   0
join_date              0
workshops_attended     0
project_score         12
github_profile        15
status                 0
email                  0
dtype: int64


In [8]:
# Percentage of missing values per column
missing_percentage = (df.isnull().sum() / len(df)) * 100
print("Percentage of Missing Values per Column:\n", missing_percentage)


Percentage of Missing Values per Column:
 member_id              0.00000
full_name              0.00000
age                    0.00000
wilaya                 6.25000
university             4.68750
level                  0.00000
track                  0.00000
role                   0.00000
join_date              0.00000
workshops_attended     0.00000
project_score          9.37500
github_profile        11.71875
status                 0.00000
email                  0.00000
dtype: float64


In [9]:
# Column names with at least one missing value
columns_with_missing = df.columns[df.isnull().any()].tolist()
print("Columns with Missing Values:", columns_with_missing)


Columns with Missing Values: ['wilaya', 'university', 'project_score', 'github_profile']


---

## Part 2: Data Cleaning

### Exo 4: Fix Data Types (15 points)

When you loaded the data, some columns were not stored in the correct type. Fix the following:

- `age` → should be **integer** (it's currently a string like `"22.0"`)
- `workshops_attended` → should be **integer** (it's currently a string)
- `join_date` → should be **datetime**

After converting, print the dtypes to confirm.

**Hint:** Convert `age` to float first, then to int.


In [10]:
# Convert age to integer
df['age'] = df['age'].astype(int)
df['age']


0      25
1      26
2      23
3      23
4      25
       ..
123    24
124    23
125    26
126    19
127    25
Name: age, Length: 128, dtype: int64

In [11]:
# Convert workshops_attended to integer
df['workshops_attended'] = df['workshops_attended'].astype(int)
df['workshops_attended']


0      7
1      4
2      7
3      2
4      0
      ..
123    3
124    4
125    0
126    0
127    3
Name: workshops_attended, Length: 128, dtype: int64

In [12]:
# Convert join_date to datetime
df['join_date'] = pd.to_datetime(df['join_date'])
df['join_date']


0     2024-04-14
1     2024-07-04
2     2024-02-27
3     2024-09-09
4     2024-07-19
         ...    
123   2024-08-11
124   2024-07-12
125   2024-07-18
126   2024-03-16
127   2024-05-11
Name: join_date, Length: 128, dtype: datetime64[us]

In [13]:
# Confirm dtypes
print(df.dtypes)


member_id                        str
full_name                        str
age                            int64
wilaya                           str
university                       str
level                            str
track                            str
role                             str
join_date             datetime64[us]
workshops_attended             int64
project_score                float64
github_profile                   str
status                           str
email                            str
dtype: object


---

### Exo 5: Handle Missing Values (10 points)

Fill missing values as follows:

- `project_score` → fill with the **median** of the column
- `wilaya` → fill with `"Unknown"`
- `university` → fill with `"Unknown"`
- `github_profile` → fill with `"Not Provided"`
- `email` → drop rows where email is missing (we can't contact them!)

After cleaning, confirm there are no more missing values.


In [14]:
# Fill missing project_score with median
median_score = df['project_score'].median()
df['project_score'] = df['project_score'].fillna(median_score)


In [15]:
# Fill missing wilaya and university with 'Unknown'
df['wilaya'] = df['wilaya'].fillna('Unknown')
df['university'] = df['university'].fillna('Unknown')


In [16]:
# Fill missing github_profile with 'Not Provided'
df['github_profile'] = df['github_profile'].fillna('Not Provided')

In [17]:
# Drop rows where email is missing
df = df.dropna(subset=['email'])


In [18]:
# Confirm no more missing values
print(df.isnull().sum())


member_id             0
full_name             0
age                   0
wilaya                0
university            0
level                 0
track                 0
role                  0
join_date             0
workshops_attended    0
project_score         0
github_profile        0
status                0
email                 0
dtype: int64


---

### Exo 6: Remove Duplicates (5 points)

1. Print how many **duplicate rows** exist in the dataset
2. Remove them, keeping the **first** occurrence
3. Print the shape of the DataFrame before and after to confirm


In [19]:
# Count duplicate rows
duplicate_count = df.duplicated().sum()
print("Number of duplicate rows:", duplicate_count)


Number of duplicate rows: 8


In [20]:
# Remove duplicates and confirm
print(df[df.duplicated])
df = df.drop_duplicates(keep='first')
print(df[df.duplicated])



        member_id       full_name  age      wilaya    university level  \
12   OMC-2024-622     Imane Saidi   26       BLIDA          UMBB    M2   
18   OMC-2024-356     Amina Saidi   20  TIZI OUZOU           ESI    M2   
21   OMC-2024-398   Sofiane Ziani   23      Béjaïa  Univ. Oran 1    M2   
61   OMC-2024-513      Amina Hadj   18       BATNA           ESI    M2   
69   OMC-2024-766    Wafa Rahmani   23      Annaba         ENSIA    M2   
81   OMC-2024-440  Amira Bensalem   26  TIZI OUZOU          UMBB    M2   
107  OMC-2024-726    Tarek Benali   26      Béjaïa          UMBB    L3   
116  OMC-2024-989     Rayan Larbi   19       Batna           ESI    L1   

                    track       role  join_date  workshops_attended  \
12              AI & Data     Member 2024-07-04                   4   
18   Software Engineering     Member 2024-07-27                   7   
21          Cybersecurity     Member 2024-02-27                   7   
61              AI & Data     Member 2024-06-20  

---

### Exo 7: Standardize Text Columns (10 points)

The `wilaya` and `status` columns have **inconsistent casing** (some values are uppercase, some lowercase, some mixed).

1. Standardize `wilaya` to **title case** (e.g. `"ALGIERS"` → `"Algiers"`)
2. Standardize `status` to **title case** (e.g. `"active"` → `"Active"`)
3. Display the unique values of each column after cleaning to confirm


In [21]:
# Standardize wilaya to title case
df['wilaya'] = df['wilaya'].str.title()



In [22]:
# Standardize status to title case
df['status'] = df['status'].str.title()



In [23]:
# Confirm unique values
print(df['wilaya'].unique())
print(df['status'].unique())


<StringArray>
[      'Sétif',       'Blida',      'Béjaïa',      'Annaba',       'Batna',
     'Tlemcen',     'Algiers', 'Constantine',  'Tizi Ouzou',        'Oran',
     'Unknown']
Length: 11, dtype: str
<StringArray>
['Active', 'Inactive']
Length: 2, dtype: str


---

## Part 3: Filtering & Selection

### Exo 8: Boolean Filtering (10 points)

Answer the following questions using boolean indexing:

1. How many members have a `project_score` **greater than 15**?
2. Show all **Core Team** or **Lead** members who are from **Algiers**
3. Show all **Active** members who attended **more than 4 workshops**


In [24]:
# 1. Members with project_score > 15
df[df['project_score'] > 15]


member_id          full_name  age       wilaya         university  \
1    OMC-2024-622        Imane Saidi   26        Blida               UMBB   
2    OMC-2024-398      Sofiane Ziani   23       Béjaïa       Univ. Oran 1   
6    OMC-2024-463     Nadine Boudiaf   25      Tlemcen              USTHB   
7    OMC-2024-317         Wafa Ouali   22        Sétif                ENP   
8    OMC-2024-274    Meriem Boudaoud   18        Sétif              USTHB   
9    OMC-2024-017        Ilyes Larbi   19      Algiers              ENSIA   
13   OMC-2024-624       Meriem Mekki   18       Annaba              ENSIA   
20   OMC-2024-181    Islem Boukhalfa   26      Algiers               UMBB   
22   OMC-2024-593    Youcef Mansouri   18      Tlemcen       Univ. Oran 1   
23   OMC-2024-922    Islem Boukhalfa   18         Oran               UMBB   
24   OMC-2024-885         Nour Larbi   20        Sétif               UDBA   
26   OMC-2024-127      Sofiane Ziani   21        Blida       Univ. Oran 1   
29   OMC-2024-037     Oussama Guerfi   19      Unknown              ENSIA   
36   OMC-2024-482       Youcef Mekki   20      Algiers               UDBA   
38   OMC-2024-720      Sofiane Larbi   21        Blida                ENP   
40   OMC-2024-676       Chaima Mekki   25  Constantine                ENP   
41   OMC-2024-225      Asma Bensalem   23        Blida              ENSIA   
42   OMC-2024-952   Oussama Mansouri   20       Annaba              USTHB   
46   OMC-2024-247        Lina Guerfi   25   Tizi Ouzou              ENSIA   
50   OMC-2024-198       Nassim Ouali   20        Blida                ESI   
52   OMC-2024-437    Walid Boukhalfa   21  Constantine  Univ. Constantine   
53   OMC-2024-595        Rima Khelif   18         Oran       Univ. Oran 1   
54   OMC-2024-991     Nassim Hadjadj   19      Tlemcen                ESI   
59   OMC-2024-684        Adem Khelif   23       Annaba       Univ. Oran 1   
65   OMC-2024-324        Rima Cherif   25   Tizi Ouzou               UMBB   
67   OMC-2024-625       Meriem Ziani   19   Tizi Ouzou                ENP   
68   OMC-2024-926         Asma Ziani   20      Tlemcen  Univ. Constantine   
70   OMC-2024-626     Nour Boukhalfa   21        Sétif               UMBB   
71   OMC-2024-892   Adem Benabdallah   19      Unknown            Unknown   
72   OMC-2024-125      Rayan Boudiaf   20        Sétif  Univ. Constantine   
75   OMC-2024-726       Tarek Benali   26       Béjaïa               UMBB   
82   OMC-2024-699        Amina Larbi   19   Tizi Ouzou                ESI   
85   OMC-2024-824    Karim Boukhalfa   26         Oran            Unknown   
90   OMC-2024-357         Wafa Ouali   25   Tizi Ouzou                ESI   
92   OMC-2024-178     Feriel Hadjadj   20         Oran            Unknown   
95   OMC-2024-896     Imane Boudaoud   24       Béjaïa  Univ. Constantine   
97   OMC-2024-981    Rayan Boukhalfa   26         Oran               UDBA   
102  OMC-2024-717      Cyrine Benali   23        Sétif                ENP   
103  OMC-2024-655       Karim Benali   19        Blida               UMBB   
104  OMC-2024-245          Sara Hadj   19      Unknown               UMBB   
105  OMC-2024-272       Sara Boudiaf   25       Béjaïa                ENP   
109  OMC-2024-066         Wafa Saidi    0   Tizi Ouzou  Univ. Constantine   
115  OMC-2024-384      Tarek Hadjadj   20       Annaba       Univ. Oran 1   
117  OMC-2024-235        Nour Khelif   25   Tizi Ouzou                ESI   
123  OMC-2024-458         Amina Hadj   24   Tizi Ouzou               UDBA   
124  OMC-2024-806       Nadine Mekki   23      Algiers       Univ. Oran 1   
127  OMC-2024-446  Ilyes Benabdallah   25      Unknown  Univ. Constantine   

    level                 track       role  join_date  workshops_attended  \
1      M2             AI & Data     Member 2024-07-04                   4   
2      M2         Cybersecurity     Member 2024-02-27                   7   
6      M2  Software Engineering     Member 2024-01-02                   7   
7  

In [25]:
# 2. Core Team or Lead members from Algiers
df[(df['role'].isin(['Core Team', 'Lead']))&(df['wilaya'] == 'Algiers')]


member_id        full_name  age   wilaya    university level  \
36   OMC-2024-482     Youcef Mekki   20  Algiers          UDBA    L2   
77   OMC-2024-084  Oussama Boudiaf   25  Algiers          UDBA    L1   
98   OMC-2024-784     Mehdi Guerfi   19  Algiers         ENSIA    M1   
124  OMC-2024-806     Nadine Mekki   23  Algiers  Univ. Oran 1    M1   

                    track       role  join_date  workshops_attended  \
36              AI & Data       Lead 2024-01-25                   0   
77   Software Engineering       Lead 2024-03-05                   0   
98        Web Development  Core Team 2024-08-09                   2   
124             AI & Data  Core Team 2024-07-12                   4   

     project_score        github_profile    status                      email  
36           19.40   github.com/youcef35    Active     youcef.mekki@email.com  
77           11.40  github.com/oussama70    Active  oussama.boudiaf@email.com  
98           13.35          Not Provided  Inactive     mehdi.guerfi@email.com  
124          17.00   github.com/nadine55    Active     nadine.mekki@email.com

In [26]:
# 3. Active members who attended more than 4 workshops
df[(df['workshops_attended'] > 4)]


member_id          full_name  age       wilaya         university  \
0    OMC-2024-465    Nadine Bensalem   25        Sétif                ESI   
2    OMC-2024-398      Sofiane Ziani   23       Béjaïa       Univ. Oran 1   
6    OMC-2024-463     Nadine Boudiaf   25      Tlemcen              USTHB   
7    OMC-2024-317         Wafa Ouali   22        Sétif                ENP   
9    OMC-2024-017        Ilyes Larbi   19      Algiers              ENSIA   
13   OMC-2024-624       Meriem Mekki   18       Annaba              ENSIA   
14   OMC-2024-902       Feriel Larbi   25       Béjaïa              USTHB   
16   OMC-2024-992    Cyrine Bensalem   26        Blida            Unknown   
17   OMC-2024-356        Amina Saidi   20   Tizi Ouzou                ESI   
20   OMC-2024-181    Islem Boukhalfa   26      Algiers               UMBB   
32   OMC-2024-730      Imane Meziane   22         Oran               UDBA   
33   OMC-2024-249       Chaima Saidi   18       Annaba                ESI   
35   OMC-2024-440     Amira Bensalem   26   Tizi Ouzou               UMBB   
39   OMC-2024-459     Youcef Meziane   23        Blida               UMBB   
41   OMC-2024-225      Asma Bensalem   23        Blida              ENSIA   
42   OMC-2024-952   Oussama Mansouri   20       Annaba              USTHB   
43   OMC-2024-091        Houda Larbi   23   Tizi Ouzou                ESI   
45   OMC-2024-389       Amira Bellal   18        Blida  Univ. Constantine   
48   OMC-2024-662  Tarek Benabdallah   19   Tizi Ouzou              USTHB   
49   OMC-2024-828     Walid Boudaoud   20      Algiers              USTHB   
50   OMC-2024-198       Nassim Ouali   20        Blida                ESI   
51   OMC-2024-223    Imane Boukhalfa  150       Annaba               UDBA   
53   OMC-2024-595        Rima Khelif   18         Oran       Univ. Oran 1   
62   OMC-2024-827   Fatima Boukhalfa   21       Béjaïa                ESI   
63   OMC-2024-416       Walid Benali   25        Blida                ENP   
65   OMC-2024-324        Rima Cherif   25   Tizi Ouzou               UMBB   
66   OMC-2024-664     Nour Boukhalfa   20        Blida       Univ. Oran 1   
68   OMC-2024-926         Asma Ziani   20      Tlemcen  Univ. Constantine   
74   OMC-2024-123         Amira Hadj   18        Sétif              ENSIA   
75   OMC-2024-726       Tarek Benali   26       Béjaïa               UMBB   
80   OMC-2024-460       Malek Benali   22         Oran              USTHB   
83   OMC-2024-792     Cyrine Meziane   19      Unknown                ESI   
90   OMC-2024-357         Wafa Ouali   25   Tizi Ouzou                ESI   
92   OMC-2024-178     Feriel Hadjadj   20         Oran            Unknown   
95   OMC-2024-896     Imane Boudaoud   24       Béjaïa  Univ. Constantine   
96   OMC-2024-638    Lotfi Boukhalfa   21      Algiers               UMBB   
100  OMC-2024-825       Houda Bellal  999        Blida                ESI   
101  OMC-2024-255        Karim Mekki   19      Unknown               UMBB   
102  OMC-2024-717      Cyrine Benali   23        Sétif                ENP   
106  OMC-2024-250     Sofiane Guerfi   22      Tlemcen       Univ. Oran 1   
110  OMC-2024-094     Rayan Boudaoud   19  Constantine              ENSIA   
112  OMC-2024-187        Rayan Larbi   19        Sétif               UDBA   
113  OMC-2024-739       Imane Khelif   25        Sétif               UDBA   
114  OMC-2024-409      Wafa Boudaoud   24   Tizi Ouzou       Univ. Oran 1   
115  OMC-2024-384      Tarek Hadjadj   20       Annaba       Univ. Oran 1   

    level                 track       role  join_date  workshops_attended  \
0      L1       Web Development  Core Team 2024-04-14                   7   
2      M2         Cybersecurity     Member 2024-02-27                   7   
6      M2  Software Engineering     Member 2024-01-02                   7   
7      M2         Cybersecurity     Member 2024-07-22                   6   
9      M1       Web Development     Member 2024-02-16                   7   
13 

---

### Exo 9: Selection with `.isin()` and `.loc[]` (10 points)

1. Use `.isin()` to filter members whose `track` is either `"AI & Data"` or `"Cybersecurity"`. How many are there?
2. Use `.loc[]` to display only the `full_name`, `track`, and `project_score` columns for **M1 and M2** level members.


In [27]:
# 1. Filter by track using .isin()
df[df['track'].isin(['AI & Data', 'Cybersecurity'])]


member_id            full_name  age       wilaya         university  \
1    OMC-2024-622          Imane Saidi   26        Blida               UMBB   
2    OMC-2024-398        Sofiane Ziani   23       Béjaïa       Univ. Oran 1   
7    OMC-2024-317           Wafa Ouali   22        Sétif                ENP   
8    OMC-2024-274      Meriem Boudaoud   18        Sétif              USTHB   
11   OMC-2024-360        Amira Rahmani   20  Constantine       Univ. Oran 1   
14   OMC-2024-902         Feriel Larbi   25       Béjaïa              USTHB   
19   OMC-2024-222         Chaima Larbi   22      Tlemcen                ENP   
22   OMC-2024-593      Youcef Mansouri   18      Tlemcen       Univ. Oran 1   
23   OMC-2024-922      Islem Boukhalfa   18         Oran               UMBB   
26   OMC-2024-127        Sofiane Ziani   21        Blida       Univ. Oran 1   
29   OMC-2024-037       Oussama Guerfi   19      Unknown              ENSIA   
31   OMC-2024-894          Rima Khelif   25        Batna               UDBA   
32   OMC-2024-730        Imane Meziane   22         Oran               UDBA   
36   OMC-2024-482         Youcef Mekki   20      Algiers               UDBA   
38   OMC-2024-720        Sofiane Larbi   21        Blida                ENP   
39   OMC-2024-459       Youcef Meziane   23        Blida               UMBB   
41   OMC-2024-225        Asma Bensalem   23        Blida              ENSIA   
49   OMC-2024-828       Walid Boudaoud   20      Algiers              USTHB   
56   OMC-2024-513           Amina Hadj   18        Batna                ESI   
59   OMC-2024-684          Adem Khelif   23       Annaba       Univ. Oran 1   
60   OMC-2024-363          Walid Saidi   21      Tlemcen              ENSIA   
62   OMC-2024-827     Fatima Boukhalfa   21       Béjaïa                ESI   
66   OMC-2024-664       Nour Boukhalfa   20        Blida       Univ. Oran 1   
72   OMC-2024-125        Rayan Boudiaf   20        Sétif  Univ. Constantine   
73   OMC-2024-765          Amine Saidi   23         Oran            Unknown   
78   OMC-2024-224  Yasmine Benabdallah   20      Tlemcen              USTHB   
80   OMC-2024-460         Malek Benali   22         Oran              USTHB   
87   OMC-2024-379           Adem Ouali   25      Tlemcen                ESI   
88   OMC-2024-813       Meriem Meziane   25        Blida               UDBA   
91   OMC-2024-047          Rayan Mekki   20        Blida              ENSIA   
94   OMC-2024-971         Bilal Cherif   25        Blida                ESI   
95   OMC-2024-896       Imane Boudaoud   24       Béjaïa  Univ. Constantine   
103  OMC-2024-655         Karim Benali   19        Blida               UMBB   
106  OMC-2024-250       Sofiane Guerfi   22      Tlemcen       Univ. Oran 1   
109  OMC-2024-066           Wafa Saidi    0   Tizi Ouzou  Univ. Constantine   
115  OMC-2024-384        Tarek Hadjadj   20       Annaba       Univ. Oran 1   
117  OMC-2024-235          Nour Khelif   25   Tizi Ouzou                ESI   
118  OMC-2024-299       Fatima Boudiaf   20        Blida                ENP   
122  OMC-2024-679       Bilal Bensalem   20      Algiers              USTHB   
124  OMC-2024-806         Nadine Mekki   23      Algiers       Univ. Oran 1   

    level          track       role  join_date  workshops_attended  \
1      M2      AI & Data     Member 2024-07-04                   4   
2      M2  Cybersecurity     Member 2024-02-27                   7   
7      M2  Cybersecurity     Member 2024-07-22                   6   
8      M1  Cybersecurity     Member 2024-08-12                   4   
11     L1  Cybersecurity  Core Team 2024-01-03                   3   
14     M2  Cybersecurity     Member 2024-04-24                   6   
19     L1  Cybersecurity     Member 2024-04-12                   2   
22     M1  Cybersecurity     Member 2024-04-04                   4   
23     L2      AI & Data     Member 2024-01-19                   3   
26     L3      AI & Data  Core Team 2024-04-16                   0  

In [28]:
# 2. M1 and M2 members — selected columns using .loc[]
df.loc[df['level'].isin(['M1', 'M2']), ['full_name', 'track','project_score']]


full_name                 track  project_score
1            Imane Saidi             AI & Data          16.10
2          Sofiane Ziani         Cybersecurity          15.40
3           Wafa Rahmani  Software Engineering          11.10
4         Nassim Rahmani       Web Development           9.00
6         Nadine Boudiaf  Software Engineering          18.20
7             Wafa Ouali         Cybersecurity          16.10
8        Meriem Boudaoud         Cybersecurity          17.80
9            Ilyes Larbi       Web Development          17.30
13          Meriem Mekki      Embedded Systems          18.10
14          Feriel Larbi         Cybersecurity          13.35
15        Meriem Meziane       Web Development          10.80
17           Amina Saidi  Software Engineering           9.30
22       Youcef Mansouri         Cybersecurity          18.20
24            Nour Larbi  Software Engineering          18.00
27        Besma Boudaoud  Software Engineering          12.40
29        Oussama Guerfi         Cybersecurity          18.80
31           Rima Khelif             AI & Data          11.40
32         Imane Meziane         Cybersecurity          12.60
35        Amira Bensalem      Embedded Systems           8.80
37      Djamila Mansouri       Web Development          11.30
39        Youcef Meziane         Cybersecurity          13.10
40          Chaima Mekki       Web Development          16.60
43           Houda Larbi       Web Development          13.00
44    Cyrine Benabdallah      Embedded Systems          12.60
52       Walid Boukhalfa  Software Engineering          17.20
56            Amina Hadj             AI & Data           8.30
57            Nour Ouali      Embedded Systems          12.90
59           Adem Khelif             AI & Data          17.20
62      Fatima Boukhalfa             AI & Data          14.90
67          Meriem Ziani      Embedded Systems          17.50
68            Asma Ziani  Software Engineering          19.60
70        Nour Boukhalfa      Embedded Systems          18.70
73           Amine Saidi         Cybersecurity          11.90
78   Yasmine Benabdallah         Cybersecurity          13.40
83        Cyrine Meziane  Software Engineering          12.30
89       Meriem Mansouri      Embedded Systems          13.50
91           Rayan Mekki             AI & Data          13.60
92        Feriel Hadjadj      Embedded Systems          18.40
98          Mehdi Guerfi       Web Development          13.35
101          Karim Mekki      Embedded Systems          11.70
104            Sara Hadj  Software Engineering          17.30
105         Sara Boudiaf      Embedded Systems          20.00
109           Wafa Saidi             AI & Data          18.30
110       Rayan Boudaoud       Web Development          14.30
111      Oussama Hadjadj       Web Development          12.90
115        Tarek Hadjadj         Cybersecurity          15.40
117          Nour Khelif         Cybersecurity          16.70
118       Fatima Boudiaf             AI & Data          13.10
119          Karim Larbi       Web Development           8.30
120       Cyrine Meziane       Web Development          11.90
121        Adem Bensalem  Software Engineering           9.40
123           Amina Hadj       Web Development          19.80
124         Nadine Mekki             AI & Data          17.00
125        Ayoub Hadjadj       Web Development          13.35
126          Wafa Khelif      Embedded Systems          13.35
127    Ilyes Benabdallah      Embedded Systems          18.00

---

### Exo 10: String Filtering (10 points)

Use `.str` methods to answer:

1. Find all members whose `full_name` **starts with the letter 'A'**
2. Find all members whose `email` **contains `benali`** (case-insensitive)
3. How many members have a `github_profile` that is **not** `"Not Provided"`?

**Hint:** For question 3, you can combine `.str` filtering or just use a comparison.


In [29]:
# 1. Names starting with 'A'
df[df['full_name'].str.startswith('A')]


member_id         full_name  age       wilaya         university  \
11   OMC-2024-360     Amira Rahmani   20  Constantine       Univ. Oran 1   
17   OMC-2024-356       Amina Saidi   20   Tizi Ouzou                ESI   
35   OMC-2024-440    Amira Bensalem   26   Tizi Ouzou               UMBB   
41   OMC-2024-225     Asma Bensalem   23        Blida              ENSIA   
45   OMC-2024-389      Amira Bellal   18        Blida  Univ. Constantine   
56   OMC-2024-513        Amina Hadj   18        Batna                ESI   
59   OMC-2024-684       Adem Khelif   23       Annaba       Univ. Oran 1   
64   OMC-2024-666       Amine Ouali   21       Béjaïa       Univ. Oran 1   
68   OMC-2024-926        Asma Ziani   20      Tlemcen  Univ. Constantine   
71   OMC-2024-892  Adem Benabdallah   19      Unknown            Unknown   
73   OMC-2024-765       Amine Saidi   23         Oran            Unknown   
74   OMC-2024-123        Amira Hadj   18        Sétif              ENSIA   
76   OMC-2024-413    Ayoub Boudaoud   25       Annaba               UMBB   
82   OMC-2024-699       Amina Larbi   19   Tizi Ouzou                ESI   
87   OMC-2024-379        Adem Ouali   25      Tlemcen                ESI   
99   OMC-2024-594      Amina Khelif   20   Tizi Ouzou              USTHB   
121  OMC-2024-721     Adem Bensalem   20       Béjaïa              USTHB   
123  OMC-2024-458        Amina Hadj   24   Tizi Ouzou               UDBA   
125  OMC-2024-004     Ayoub Hadjadj   26        Blida              ENSIA   

    level                 track       role  join_date  workshops_attended  \
11     L1         Cybersecurity  Core Team 2024-01-03                   3   
17     M2  Software Engineering     Member 2024-07-27                   7   
35     M2      Embedded Systems     Member 2024-02-17                   5   
41     L1             AI & Data     Member 2024-08-02                   5   
45     L3      Embedded Systems  Core Team 2024-09-22                   7   
56     M2             AI & Data     Member 2024-06-20                   4   
59     M2             AI & Data     Member 2024-04-10                   3   
64     L3      Embedded Systems       Lead 2024-09-02                   1   
68     M1  Software Engineering       Lead 2024-07-17                   6   
71     L2  Software Engineering       Lead 2024-03-18                   0   
73     M1         Cybersecurity  Core Team 2024-05-01                   0   
74     L2      Embedded Systems  Core Team 2024-04-17                   5   
76     L2       Web Development     Member 2024-03-14                   2   
82     L3      Embedded Systems     Member 2024-08-01                   4   
87     L1         Cybersecurity     Member 2024-01-22                   1   
99     L1       Web Development     Member 2024-09-28                   2   
121    M2  Software Engineering  Core Team 2024-02-05                   4   
123    M2       Web Development     Member 2024-08-11                   3   
125    M1       Web Development  Core Team 2024-07-18                   0   

     project_score      github_profile    status                     email  
11           12.50        Not Provided    Active   amira.rahmani@email.com  
17            9.30  github.com/amina56  Inactive     amina.saidi@email.com  
35            8.80  github.com/amira77    Active  amira.bensalem@email.com  
41           16.80        Not Provided    Active   asma.bensalem@email.com  
45            9.90  github.com/amira47    Active    amira.bellal@email.com  
56            8.30  github.com/amina39    Active      amina.hadj@email.com  
59           17.20   github.com/adem42    Active     adem.khelif@email.com  
64           10.80  github.com/amine97    Active     amine.ouali@email.com  
68           19.60   github.com/asma35    Active      asma.ziani@email.com  
71           15.20   github.com/adem72    Active         invalid@email.net  
73           11.90  github.com/amine17    Active     amine.saidi@email.com  
74           13.35  gi

In [30]:
# 2. Emails containing 'benali'
df[df['email'].str.contains('benali')]


member_id      full_name  age  wilaya university level  \
5    OMC-2024-808  Feriel Benali   23   Blida      ENSIA    L2   
63   OMC-2024-416   Walid Benali   25   Blida        ENP    L2   
75   OMC-2024-726   Tarek Benali   26  Béjaïa       UMBB    L3   
80   OMC-2024-460   Malek Benali   22    Oran      USTHB    L3   
102  OMC-2024-717  Cyrine Benali   23   Sétif        ENP    L1   
103  OMC-2024-655   Karim Benali   19   Blida       UMBB    L2   

                track       role  join_date  workshops_attended  \
5     Web Development     Member 2024-03-04                   0   
63   Embedded Systems       Lead 2024-06-15                   7   
75   Embedded Systems     Member 2024-06-20                   5   
80      Cybersecurity       Lead 2024-03-18                   6   
102   Web Development       Lead 2024-01-20                   7   
103     Cybersecurity  Core Team 2024-02-22                   1   

     project_score       github_profile  status                    email  
5             9.60  github.com/feriel55  Active  feriel.benali@email.com  
63           13.30   github.com/walid37  Active   walid.benali@email.com  
75           16.00   github.com/tarek25  Active   tarek.benali@email.com  
80           13.35   github.com/malek46  Active   malek.benali@email.com  
102          18.00  github.com/cyrine46  Active  cyrine.benali@email.com  
103          15.10   github.com/karim14  Active   karim.benali@email.com

In [31]:
# 3. Members with a real GitHub profile
df[df['github_profile'] != 'Not Provided']


member_id          full_name  age      wilaya         university  \
0    OMC-2024-465    Nadine Bensalem   25       Sétif                ESI   
1    OMC-2024-622        Imane Saidi   26       Blida               UMBB   
2    OMC-2024-398      Sofiane Ziani   23      Béjaïa       Univ. Oran 1   
4    OMC-2024-345     Nassim Rahmani   25       Batna              ENSIA   
5    OMC-2024-808      Feriel Benali   23       Blida              ENSIA   
..            ...                ...  ...         ...                ...   
120  OMC-2024-378     Cyrine Meziane   20        Oran       Univ. Oran 1   
121  OMC-2024-721      Adem Bensalem   20      Béjaïa              USTHB   
124  OMC-2024-806       Nadine Mekki   23     Algiers       Univ. Oran 1   
126  OMC-2024-240        Wafa Khelif   19  Tizi Ouzou  Univ. Constantine   
127  OMC-2024-446  Ilyes Benabdallah   25     Unknown  Univ. Constantine   

    level                 track       role  join_date  workshops_attended  \
0      L1       Web Development  Core Team 2024-04-14                   7   
1      M2             AI & Data     Member 2024-07-04                   4   
2      M2         Cybersecurity     Member 2024-02-27                   7   
4      M2       Web Development     Member 2024-07-19                   0   
5      L2       Web Development     Member 2024-03-04                   0   
..    ...                   ...        ...        ...                 ...   
120    M1       Web Development  Core Team 2024-01-15                   3   
121    M2  Software Engineering  Core Team 2024-02-05                   4   
124    M1             AI & Data  Core Team 2024-07-12                   4   
126    M1      Embedded Systems  Core Team 2024-03-16                   0   
127    M1      Embedded Systems       Lead 2024-05-11                   3   

     project_score        github_profile    status  \
0            13.35   github.com/nadine62    Active   
1            16.10    github.com/imane85  Inactive   
2            15.40  github.com/sofiane45    Active   
4             9.00   github.com/nassim64    Active   
5             9.60   github.com/feriel55    Active   
..             ...                   ...       ...   
120          11.90   github.com/cyrine70    Active   
121           9.40     github.com/adem23    Active   
124          17.00   github.com/nadine55    Active   
126          13.35     github.com/wafa85    Active   
127          18.00    github.com/ilyes21    Active   

                           email  
0      nadine.bensalem@email.com  
1          imane.saidi@email.com  
2        sofiane.ziani@email.com  
4       nassim.rahmani@email.com  
5        feriel.benali@email.com  
..                           ...  
120     cyrine.meziane@email.com  
121                 @nodomain.dz  
124       nadine.mekki@email.com  
126        wafa.khelif@email.com  
127  ilyes.benabdallah@email.com  

[106 rows x 14 columns]

---

## Part 4: Exploratory Analysis

### Exo 11: Value Counts & Distributions (10 points)

1. Which **3 wilayas** have the most members? Display the count for each.
2. What is the distribution of members across **tracks**? Show as percentages (normalized).
3. How many members are **Active** vs **Inactive**?


In [34]:
# 1. Top 3 wilayas by member count
top_wilayas = df['wilaya'].value_counts().sort_values(ascending=False).head(3)
print("Top 3 Wilayas by Member Count:\n", top_wilayas)


Top 3 Wilayas by Member Count:
 wilaya
Blida         21
Tizi Ouzou    15
Sétif         14
Name: count, dtype: int64


In [35]:
# 2. Track distribution as percentages
df['track'].value_counts(normalize=True) * 100

track
Embedded Systems        24.166667
Web Development         23.333333
Cybersecurity           20.000000
Software Engineering    19.166667
AI & Data               13.333333
Name: proportion, dtype: float64

In [36]:
# 3. Active vs Inactive count
df['status'].value_counts()


status
Active      102
Inactive     18
Name: count, dtype: int64

---

### Exo 12: Group-Level Statistics (10 points)

Use `.groupby()` to answer the following:

1. What is the **average `project_score`** per `track`? Sort from highest to lowest.
2. What is the **average number of workshops attended** per `level` (L1, L2, L3, M1, M2)?
3. Which `role` has the **highest median `project_score`**?


In [37]:
# 1. Average project_score per track
df.groupby('track')['project_score'].mean().sort_values(ascending=False)


track
AI & Data               15.153125
Embedded Systems        14.456897
Software Engineering    13.934783
Cybersecurity           13.904167
Web Development         13.164286
Name: project_score, dtype: float64

In [38]:
# 2. Average workshops attended per level
df.groupby('level')['workshops_attended'].mean().sort_values(ascending=False)


level
L3    4.105263
L1    3.772727
M2    3.642857
M1    3.214286
L2    2.652174
Name: workshops_attended, dtype: float64

In [39]:
# 3. Role with highest median project_score
df.groupby('role')['project_score'].median().sort_values(ascending=False).head(1)


role
Lead    15.2
Name: project_score, dtype: float64

---

## Part 5: Bonus — Optional

### Bonus 1: Email Validator (3 points)

Remember the email validator you wrote in **Checkpoint 1.1**? Time to bring it back — but this time inside a Pandas DataFrame.

The `email` column contains some invalid entries that slipped through. Use the **same regex rules** from Checkpoint 1.1:

- Contains exactly one `@`
- Has characters before and after `@`
- Only contains letters, numbers, dots, and underscores before `@`
- Ends with `.com`, `.dz`, or `.edu`

Your tasks:
1. Write the function `is_valid_email(email)` using `re`
2. Apply it to the `email` column to create a new boolean column called `email_valid`
3. Print how many emails are **valid** and how many are **invalid**
4. Display the rows where the email is **invalid** — show only `full_name` and `email`

**Hint:** Use `.apply()` to apply your function across the column.


In [40]:
import re

def is_valid_email(email):
    pattern = r'^[\w\.]+@[\w]+\.(com|dz|edu)$'
    return bool(re.match(pattern, email))


In [41]:
# Apply the function to create the email_valid column
df['email_valid'] = df['email'].apply(is_valid_email)

In [42]:
# Count valid vs invalid
print(df['email_valid'].value_counts())


email_valid
True     105
False     15
Name: count, dtype: int64


In [43]:
# Display invalid rows (full_name and email only)
df[~df['email_valid']][['full_name', 'email']]


full_name                   email
8      Meriem Boudaoud       double@@email.com
14        Feriel Larbi          no_at_sign.com
26       Sofiane Ziani     spaces in@email.com
39      Youcef Meziane       double@@email.com
48   Tarek Benabdallah            @nodomain.dz
49      Walid Boudaoud         missing_domain@
57          Nour Ouali  special!char@email.com
71    Adem Benabdallah       invalid@email.net
96     Lotfi Boukhalfa       invalid@email.org
97     Rayan Boukhalfa     spaces in@email.com
109         Wafa Saidi         missing_domain@
113       Imane Khelif          user@@email.dz
115      Tarek Hadjadj            no_tld@email
121      Adem Bensalem            @nodomain.dz
122     Bilal Bensalem          no_at_sign.com

---

### Bonus 2: Detect & Handle Outliers (4 points)

The `age` column has some clearly invalid values (e.g. someone is 150 years old).

1. Display the rows where `age` is **outside the range [17, 30]**
2. Replace those invalid ages with `NaN` using `.loc[]`
3. Fill the NaN ages with the **median** age
4. Confirm the min and max age are now within valid range


In [44]:
# 1. Display outlier rows
df[(df['age']<17) | (df['age']>30)]


member_id        full_name  age      wilaya         university level  \
51   OMC-2024-223  Imane Boukhalfa  150      Annaba               UDBA    L1   
100  OMC-2024-825     Houda Bellal  999       Blida                ESI    L2   
109  OMC-2024-066       Wafa Saidi    0  Tizi Ouzou  Univ. Constantine    M1   

                    track       role  join_date  workshops_attended  \
51   Software Engineering  Core Team 2024-06-02                   6   
100       Web Development  Core Team 2024-06-07                   7   
109             AI & Data     Member 2024-07-13                   0   

     project_score      github_profile    status                      email  \
51            11.3  github.com/imane25  Inactive  imane.boukhalfa@email.com   
100           12.7  github.com/houda92  Inactive     houda.bellal@email.com   
109           18.3   github.com/wafa83  Inactive            missing_domain@   

     email_valid  
51          True  
100         True  
109        False

In [45]:
# 2. Replace invalid ages with NaN
df.loc[(df['age']<17) | (df['age']>30), 'age'] = np.nan


In [46]:
# 3. Fill NaN ages with median
median_age = df['age'].median()
df['age'] = df['age'].fillna(median_age)


In [47]:
# 4. Confirm valid range
print('minimum age:', df['age'].min(), '\nmaximum age:', df['age'].max())


minimum age: 18.0 
maximum age: 26.0


---

### Bonus 3: Top Performers (3 points)

Create a new DataFrame called `top_performers` that contains only members who satisfy **all** of the following:

- `status` is `"Active"`
- `project_score` is in the **top 10%** of all scores
- `workshops_attended` is **greater than or equal to 5**

Display their `full_name`, `track`, `project_score`, and `workshops_attended`, sorted by `project_score` descending.

**Hint:** Use `df['project_score'].quantile(0.9)` to get the 90th percentile.


In [48]:
# Build top_performers DataFrame
df.loc[(df['status'] == 'Active') & (df['project_score'] > df['project_score'].quantile(0.9)) & (df['workshops_attended'] >= 5), ['full_name', 'track', 'project_score', 'workshops_attended','project_score']]


full_name                 track  project_score  workshops_attended  \
20  Islem Boukhalfa      Embedded Systems           19.4                   5   
50     Nassim Ouali  Software Engineering           19.3                   5   
53      Rima Khelif  Software Engineering           18.8                   6   
65      Rima Cherif      Embedded Systems           19.1                   5   
68       Asma Ziani  Software Engineering           19.6                   6   
92   Feriel Hadjadj      Embedded Systems           18.4                   5   

    project_score  
20           19.4  
50           19.3  
53           18.8  
65           19.1  
68           19.6  
92           18.4

---

## Acknowledgments

- Notebook authored by: Open Minds Club - AI Leadership
- Dataset: Synthetic OMC member data generated by Claude for educational purposes
- Workshop content inspired by Pandas official documentation: https://pandas.pydata.org/docs/